### PR TITLE
Introduce Basic interfaces and skeleton for Bulk PubSub

### PR DESCRIPTION
**Pull request by DeepanshuA**
# Description

Introduces basic interfaces and skeleton for Batch PubSub.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

related to: https://github.com/dapr/dapr/issues/2218

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pubsub/pubsub.go
Comment (shubham1172) [inline pubsub/pubsub.go:33]: Comment for all public types/functions/methods

File: pubsub/pubsub.go
Comment (shubham1172) [inline pubsub/pubsub.go:42]: Should `BatchHandler` also return more data with error?

File: pubsub/pubsub.go
Comment (shubham1172) [inline pubsub/pubsub.go:61]:
```suggestion
	return DefaultBatcher{
		p: pubsub,
	}
```

File: pubsub/pubsub.go
Comment (shubham1172) [inline pubsub/pubsub.go:57]:
```suggestion
func NewDefaultBatcher(pubsub PubSub) DefaultBatcher {
```

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:31]: Is this required now?

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:53]: Is this required now?

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:58]:
```suggestion
	ID     string            `json:id`
```

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:57]:
```suggestion
type BatchMessage struct {
```

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:48]: nit: first line of comment to contain the type name https://go.dev/doc/comment, missing comment for `BatchSubscribeConfig`

File: pubsub/responses.go
Comment (shubham1172) [inline pubsub/responses.go:53]: nit: missing comments

File: pubsub/responses.go
Comment (shubham1172) [inline pubsub/responses.go:28]:
```suggestion
	// PublishSuccess means that the message was published successfully.
```
Same for fail.

File: pubsub/responses.go
Comment (shubham1172) [inline pubsub/responses.go:39]:
```suggestion
type AppBatchResponseItem struct {
```

File: pubsub/pubsub.go
Comment (DeepanshuA) [reply pubsub/pubsub.go:42]: Yeah, it will return.

File: pubsub/pubsub.go
Comment (DeepanshuA) [reply pubsub/pubsub.go:33]: Still WIP PR. Working on docs only.

File: pubsub/requests.go
Comment (DeepanshuA) [reply pubsub/requests.go:48]: https://go.dev/doc/comment is helpful link.

File: pubsub/pubsub.go
Comment (shubham1172) [inline pubsub/pubsub.go:33]: nit: docs

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:31]: ContentType not needed, right?

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:27]:
```suggestion
	Entries        []BulkMessageEntry `json:"data"`
```

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:48]:
```suggestion
// BulkMessage represents bulk message arriving from a message bus instance.
```

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:55]:
```suggestion
// BulkMessageEntry represents a single message inside a bulk request.
```

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:64]:
```suggestion
// BulkSubscribeConfig contains bulk subscribe related configuration. It depends on specific components to support these.
```

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:67]:
```suggestion
	MaxBulkLatencyMilliSeconds int `json:"maxBulkLatencyMilliSeconds"`
```

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:68]:
```suggestion
	MaxBulkSizeBytes      int `json:"maxBulkSizeBytes"`
```

File: pubsub/pubsub.go
Comment (shubham1172) [inline pubsub/pubsub.go:67]:
```suggestion
	return DefaultBatcher{
		p: pubsub,
	}
```

File: pubsub/pubsub.go
Comment (shubham1172) [inline pubsub/pubsub.go:33]:
```suggestion
// BulkMessager is implemented by any pubsub component that has BulkPublish and BulkSubscribe methods. 
```

similar to https://pkg.go.dev/fmt#Stringer

**Review by shubham1172 — APPROVED** [review]

File: pubsub/pubsub.go
Comment (DeepanshuA) [inline pubsub/pubsub.go:47]: The return type here is wrong - it should be an array of BulkSubscriberesponseEntry or simply BulkSubscribeResponse. Will change it - just testing few things out.

File: pubsub/pubsub.go
Comment (DeepanshuA) [inline pubsub/pubsub.go:36]: Return type be error only - BulkHandler to indicate correct status for contained items in bulk.

File: pubsub/pubsub.go
Comment (mukundansundar) [inline pubsub/pubsub.go:72]: should this be unimplemented for now in this PR?

File: pubsub/pubsub.go
Comment (mukundansundar) [inline pubsub/pubsub.go:78]: should this return an unimplemented error for now in this PR? And have a follow up PR to complete the implementation?

**Review by mukundansundar — COMMENTED** [review]
@DeepanshuA @shubham1172 thoughts?

File: pubsub/pubsub.go
Comment (shubham1172) [reply pubsub/pubsub.go:72]: Yes, I will create a follow-up PR after this.

File: pubsub/pubsub.go
Comment (shubham1172) [reply pubsub/pubsub.go:78]: I am okay with either.

Comment (DeepanshuA) [reply pubsub/pubsub.go:78]: I too am - can change error conveying unimplemented

File: pubsub/pubsub.go
Comment (DeepanshuA) [reply pubsub/pubsub.go:72]: can change error conveying unimplemented

File: pubsub/pubsub.go
Comment (ItalyPaleAle) [inline pubsub/pubsub.go:34]: Should this be "BulkMessenger"? Or "BulkPublisher"?

File: pubsub/pubsub.go
Comment (ItalyPaleAle) [inline pubsub/pubsub.go:64]: Why do we need this and then embed this in every component? Couldn't the runtime do a runtime check if the component implements the `BulkMessager` interface?

Comment (mukundansundar) [reply pubsub/pubsub.go:64]: One of the idea was to have a default bulk message implementation which provides a small perf improvement based on one single request from app to runtime even though it might be split into multiple requests between runtime and broker. These are for components that do not implement bulk publish interface. 

Similarly you have a case made for subscribe as well. That is why this interface and default implementation is done here. 
@ItalyPaleAle

File: pubsub/pubsub.go
Comment (ItalyPaleAle) [reply pubsub/pubsub.go:34]: Ok. How about "Messenger" tho?

Comment (mukundansundar) [reply pubsub/pubsub.go:34]: I am OK with that. So`BulkMessenger` interface?

File: pubsub/pubsub.go
Comment (mukundansundar) [reply pubsub/pubsub.go:64]: This has been defined here similar to the bulk state apis..

Comment (ItalyPaleAle) [reply pubsub/pubsub.go:64]: I will leave it to a maintainer to decide. Personally, I think this would be best in the runtime, but I see the precedent.

File: pubsub/pubsub.go
Comment (shubham1172) [inline pubsub/pubsub.go:32]:
```suggestion
// BulkPublisher is the interface that wraps the BulkPublish method.
//
// BulkPublish publishes a collection of messages in a BulkPublishRequest to a pubsub topic
// and returns a BulkPublishResponse with individual statuses for each message.
```

File: pubsub/pubsub.go
Comment (shubham1172) [inline pubsub/pubsub.go:34]: We can remove this. Example from stdlib https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/io/io.go;l=96

File: pubsub/pubsub.go
Comment (shubham1172) [inline pubsub/pubsub.go:43]: - This is an implementation detail that should not be here (interface is agnostic).
- `BulkSubscribe` docs to be a part of `BulkSubscriber` (similar to above)

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:47]:
```suggestion
// BulkMessage represents bulk message arriving from a message bus instance.
```

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:54]:
```suggestion
// BulkMessageEntry represents a single message inside a bulk request.
```

File: pubsub/requests.go
Comment (shubham1172) [inline pubsub/requests.go:66]:
```suggestion
	MaxBulkAwaitDurationMilliSeconds int `json:"maxBulkAwaitDurationMilliSeconds"`
```
This is similar to [RocketMQ](https://www.alibabacloud.com/help/en/message-queue-for-apache-rocketmq/latest/batch-consumption), and it sounds more intuitive than latency.

File: pubsub/responses.go
Comment (shubham1172) [inline pubsub/responses.go:29]:
```suggestion
	// PublishSucceeded means that message was published successfully.
```
Only for consistency.

File: pubsub/responses.go
Comment (shubham1172) [inline pubsub/responses.go:31]:
```suggestion
	// PublishFailed represents that message publishing failed.
```

File: pubsub/responses.go
Comment (shubham1172) [inline pubsub/responses.go:41]:
```suggestion
// AppBulkResponseEntry contains status for a single message from a BulkSubscribe event.
// The subscribing app sends this to Dapr as a part of AppBulkResponse.
```

File: pubsub/pubsub.go
Comment (DeepanshuA) [reply pubsub/pubsub.go]: It is taken care of. No more mandatory for each pubsub to implement bulk pubsub APIs and runtime will check if component implements, else call default implementation - default is part of respective bulk publish and bulk subscribe PRs. Also, two different interfaces to let BulkPublish and BulkSubscribe remain independent.

File: pubsub/pubsub.go
Comment (DeepanshuA) [reply pubsub/pubsub.go]: Segregated in two different interfaces to let BulkPublish and BulkSubscribe remain independent.

File: pubsub/pubsub.go
Comment (shubham1172) [inline pubsub/pubsub.go:33]:
```suggestion
//
```

**Comment by mukundansundar** [comment]
Closing this as Skelton is already defined and merged as part of Kafka publish impl ....

**Pull request closed without merging**
